File: runtime/python/prompty/prompty/providers/openai/processor.py
"""OpenAI response processor — extracts clean results from raw LLM responses.

Handles ChatCompletion, Completion, embedding responses, and streaming.
Registered as ``openai`` in ``prompty.processors``.

Also provides shared processing logic used by the Azure processor.
"""

from __future__ import annotations

import json
from collections.abc import AsyncIterator, Iterator
from dataclasses import dataclass
from typing import Any

from ...model import Prompty
from ...tracing.tracer import trace

__all__ = ["OpenAIProcessor", "ToolCall"]


@dataclass
class ToolCall:
    """Represents a tool call extracted from an LLM response."""

    id: str
    name: str
    arguments: str


class OpenAIProcessor:
    """Processor for OpenAI responses.

    Extracts content, tool calls, or embeddings from raw API responses.
    Registered as ``openai`` in ``prompty.processors``.
    """

    @trace
    def process(
        self,
        agent: Prompty,
        response: Any,
    ) -> Any:
        return _process_response(response, agent)

    @trace
    async def process_async(
        self,
        agent: Prompty,
        response: Any,
    ) -> Any:
        return _process_response(response, agent)


# ---------------------------------------------------------------------------
# Shared extraction logic (also used by Azure processor)
# ---------------------------------------------------------------------------


def _process_response(response: Any, agent: Prompty | None = None) -> Any:
    """Extract clean result from a raw LLM response object.

    Supports:
    - ``ChatCompletion`` → content string or list of ToolCall
    - ``Completion`` → text string
    - ``CreateEmbeddingResponse`` → embedding vector(s)
    - ``ImagesResponse`` → URL or base64 data
    - ``Response`` (Responses API) → text, tool calls, or structured output
    - Iterators (streaming) → generator of content chunks
    - Passthrough for unknown types

    When *agent* has ``outputs`` with properties, a string result
    from a ``ChatCompletion`` or ``Response`` is automatically JSON-parsed.
    """
    # Import response types lazily to keep the module importable
    # even without openai installed
    try:
        from openai.types.chat.chat_completion import ChatCompletion
        from openai.types.completion import Completion
        from openai.types.create_embedding_response import CreateEmbeddingResponse
        from openai.types.images_response import ImagesResponse
    except ImportError:
        # If openai isn't installed, just pass through
        return response

    # Try importing Responses API type (available in newer SDK versions)
    try:
        from openai.types.responses import Response as ResponsesResponse
    except ImportError:
        ResponsesResponse = None  # type: ignore[assignment, misc]

    if isinstance(response, ChatCompletion):
        result = _process_chat_completion(response)
        # JSON-parse structured output when outputs is defined
        if agent is not None and agent.outputs and isinstance(result, str):
            try:
                result = json.loads(result)
            except json.JSONDecodeError:
                pass  # Fall back to raw string
        return result
    elif isinstance(response, Completion):
        return response.choices[0].text
    elif isinstance(response, CreateEmbeddingResponse):
        return _process_embedding(response)
    elif isinstance(response, ImagesResponse):
        return _process_image(response)
    elif ResponsesResponse is not None and isinstance(response, ResponsesResponse):
        return _process_responses_api(response, agent)
    else:
        # Also check by duck-typing for Responses API (object == "response")
        if hasattr(response, "object") and response.object == "response" and hasattr(response, "output"):
            return _process_responses_api(response, agent)
        # Check for streaming iterators (sync and async)
        if isinstance(response, Iterator):
            return _stream_generator(response)
        if isinstance(response, AsyncIterator):
            return _async_stream_generator(response)
        return response


def _process_chat_completion(response: Any) -> Any:
    """Extract from a ChatCompletion response."""
    choice = response.choices[0]
    message = choice.message

    # Tool calls
    if message.tool_calls:
        return [
            ToolCall(
                id=tc.id,
                name=tc.function.name,
                arguments=tc.function.arguments,
            )
            for tc in message.tool_calls
        ]

    # Refusal — when content is null but the model refused
    refusal = getattr(message, "refusal", None)
    if message.content is None and isinstance(refusal, str):
        return refusal

    # Refusal — when content is null but the model refused
    refusal = getattr(message, "refusal", None)
    if message.content is None and isinstance(refusal, str):
        return refusal

    return message.content


def _process_embedding(response: Any) -> Any:
    """Extract from an embedding response."""
    if len(response.data) == 0:
        raise ValueError("Empty embedding response")
    elif len(response.data) == 1:
        return response.data[0].embedding
    else:
        return [item.embedding for item in response.data]


def _process_image(response: Any) -> Any:
    """Extract from an image generation response."""
    if not response.data:
        raise ValueError("Empty image response")
    elif len(response.data) == 1:
        return response.data[0].url or response.data[0].b64_json
    else:
        return [d.url or d.b64_json for d in response.data]


def _process_responses_api(response: Any, agent: Prompty | None = None) -> Any:
    """Extract from a Responses API response.

    Handles:
    - Text content from ``output_text`` or output message items
    - Function tool calls from ``function_call`` output items
    - JSON-parsed content when ``agent.outputs`` is defined
    - Error responses
    """
    # Check for errors
    if hasattr(response, "error") and response.error:
        error = response.error
        msg = getattr(error, "message", str(error))
        raise ValueError(f"Responses API error: {msg}")

    output = getattr(response, "output", []) or []

    # Collect function calls
    func_calls: list[ToolCall] = []
    for item in output:
        item_type = getattr(item, "type", None)
        if item_type == "function_call":
            func_calls.append(
                ToolCall(
                    id=getattr(item, "call_id", None) or getattr(item, "id", None) or "",
                    name=getattr(item, "name", None) or "",
                    arguments=getattr(item, "arguments", None) or "",
                )
            )

    if func_calls:
        return func_calls

    # Text content — use output_text convenience field
    output_text = getattr(response, "output_text", None)
    if output_text is not None:
        if agent is not None and agent.outputs and isinstance(output_text, str):
            try:
                return json.loads(output_text)
            except json.JSONDecodeError:
                pass
        return output_text

    # Fallback: extract from output message items
    texts: list[str] = []
    for item in output:
        item_type = getattr(item, "type", None)
        if item_type == "message":
            content = getattr(item, "content", []) or []
            for part in content:
                part_type = getattr(part, "type", None)
                if part_type in ("output_text", "text"):
                    texts.append(getattr(part, "text", ""))

    if texts:
        text = "".join(texts)
        if agent is not None and agent.outputs and isinstance(text, str):
            try:
                return json.loads(text)
            except json.JSONDecodeError:
                pass
        return text

    return response


def _stream_generator(response: Any) -> Iterator[str | ToolCall]:
    """Yield content chunks, tool calls, or refusals from a streaming response.

    Handles three types of streaming deltas:
    - ``delta.content`` — yields content strings
    - ``delta.tool_calls`` — accumulates partial tool call chunks,
      yields ``ToolCall`` objects when the stream ends
    - ``delta.refusal`` — raises ``ValueError`` with the refusal message
    """
    tool_call_acc: dict[int, dict[str, str]] = {}

    for chunk in response:
        if not hasattr(chunk, "choices") or not chunk.choices:
            continue

        delta = chunk.choices[0].delta

        # Content
        if hasattr(delta, "content") and delta.content is not None:
            yield delta.content

        # Tool call deltas — accumulate index-keyed partial chunks
        if hasattr(delta, "tool_calls") and delta.tool_calls:
            for tc_delta in delta.tool_calls:
                idx = tc_delta.index
                if idx not in tool_call_acc:
                    tool_call_acc[idx] = {"id": "", "name": "", "arguments": ""}
                if tc_delta.id:
                    tool_call_acc[idx]["id"] = tc_delta.id
                if hasattr(tc_delta, "function") and tc_delta.function:
                    if tc_delta.function.name:
                        tool_call_acc[idx]["name"] = tc_delta.function.name
                    if tc_delta.function.arguments:
                        tool_call_acc[idx]["arguments"] += tc_delta.function.arguments

        # Refusal
        if hasattr(delta, "refusal") and delta.refusal is not None:
            raise ValueError(f"Model refused: {delta.refusal}")

    # Yield accumulated tool calls at the end of the stream
    for idx in sorted(tool_call_acc):
        tc = tool_call_acc[idx]
        yield ToolCall(id=tc["id"], name=tc["name"], arguments=tc["arguments"])


async def _async_stream_generator(response: Any) -> AsyncIterator[str | ToolCall]:
    """Yield content chunks, tool calls, or refusals from an async streaming response.

    Async variant of :func:`_stream_generator`.
    """
    tool_call_acc: dict[int, dict[str, str]] = {}

    async for chunk in response:
        if not hasattr(chunk, "choices") or not chunk.choices:
            continue

        delta = chunk.choices[0].delta

        if hasattr(delta, "content") and delta.content is not None:
            yield delta.content

        if hasattr(delta, "tool_calls") and delta.tool_calls:
            for tc_delta in delta.tool_calls:
                idx = tc_delta.index
                if idx not in tool_call_acc:
                    tool_call_acc[idx] = {"id": "", "name": "", "arguments": ""}
                if tc_delta.id:
                    tool_call_acc[idx]["id"] = tc_delta.id
                if hasattr(tc_delta, "function") and tc_delta.function:
                    if tc_delta.function.name:
                        tool_call_acc[idx]["name"] = tc_delta.function.name
                    if tc_delta.function.arguments:
                        tool_call_acc[idx]["arguments"] += tc_delta.function.arguments

        if hasattr(delta, "refusal") and delta.refusal is not None:
            raise ValueError(f"Model refused: {delta.refusal}")

    for idx in sorted(tool_call_acc):
        tc = tool_call_acc[idx]
        yield ToolCall(id=tc["id"], name=tc["name"], arguments=tc["arguments"])

File: runtime/typescript/packages/openai/src/processor.ts
/**
 * OpenAI processor — extracts clean results from raw OpenAI responses.
 *
 * Handles ChatCompletion, Embedding, Image, and streaming responses.
 *
 * @module
 */

import type { Prompty } from "@prompty/core";
import type { Processor } from "@prompty/core";
import type { ToolCall } from "@prompty/core";
import {
  ErrorChunk,
  InvocationUsage,
  StreamChunk,
  TextChunk,
  ToolChunk,
  UsageChunk,
  traceSpan,
} from "@prompty/core";
import { createStructuredResult } from "@prompty/core";

export class OpenAIProcessor implements Processor {
  async process(agent: Prompty, response: unknown): Promise<unknown> {
    return traceSpan("OpenAIProcessor", async (emit) => {
      emit("signature", "prompty.openai.processor.OpenAIProcessor.invoke");
      emit("inputs", { data: response });
      const result = processResponse(agent, response);
      // Don't emit result for streaming — it's a generator, not a value
      if (!isAsyncIterable(response)) {
        emit("result", result);
      }
      return result;
    });
  }

  processStream(response: AsyncIterable<unknown>): AsyncIterable<StreamChunk> {
    return processStream(response);
  }
}

/**
 * Extract clean content from an OpenAI response.
 */
export function processResponse(agent: Prompty, response: unknown): unknown {
  if (typeof response !== "object" || response === null) return response;

  // Streaming response — return content-extracting async generator
  if (isAsyncIterable(response)) {
    return legacyStreamGenerator(processStream(response));
  }

  const r = response as Record<string, unknown>;

  // Responses API — has output[] and object === "response"
  if (r.object === "response" && Array.isArray(r.output)) {
    return processResponsesApi(agent, r);
  }

  // ChatCompletion
  if (r.choices) {
    return processChatCompletion(agent, r);
  }

  // Embedding response
  if (r.data && r.object === "list") {
    return processEmbedding(r);
  }

  // Image response
  if (r.data && Array.isArray(r.data)) {
    const data = r.data as Record<string, unknown>[];
    if (data.length > 0 && ("url" in data[0] || "b64_json" in data[0])) {
      return processImage(r);
    }
  }

  return response;
}

// ---------------------------------------------------------------------------
// Streaming
// ---------------------------------------------------------------------------

/** Type guard for async iterables (PromptyStream or raw SDK stream). */
function isAsyncIterable(value: unknown): value is AsyncIterable<unknown> {
  return (
    typeof value === "object" &&
    value !== null &&
    Symbol.asyncIterator in value
  );
}

/**
 * Yield content chunks, tool calls, or refusals from a streaming response.
 *
 * Handles three types of streaming deltas:
 * - `delta.content` — yields content strings
 * - `delta.tool_calls` — accumulates partial tool call chunks,
 *   yields ToolCall objects when the stream ends
 * - `delta.refusal` — throws Error with the refusal message
 *
 * Matches the Python `_stream_generator` / `_async_stream_generator`.
 */
export async function* processStream(
  response: AsyncIterable<unknown>,
): AsyncGenerator<StreamChunk> {
  const toolCallAcc: Map<number, { id: string; name: string; arguments: string }> = new Map();
  let usage: InvocationUsage | undefined;

  try {
    for await (const chunk of response) {
      const c = chunk as Record<string, unknown>;
      const error = c.error as Record<string, unknown> | undefined;
      if (error) {
        yield new ErrorChunk({
          message: typeof error.message === "string" ? error.message : "OpenAI stream failed",
        });
        return;
      }

      const eventType = c.type as string | undefined;
      if (eventType === "response.output_text.delta" && typeof c.delta === "string" && c.delta) {
        yield new TextChunk({ value: c.delta });
      } else if (
        (eventType === "response.output_item.added" || eventType === "response.output_item.done") &&
        isRecord(c.item) &&
        c.item.type === "function_call"
      ) {
        const idx = typeof c.output_index === "number" ? c.output_index : toolCallAcc.size;
        toolCallAcc.set(idx, {
          id: stringValue(c.item.call_id ?? c.item.id),
          name: stringValue(c.item.name),
          arguments: stringValue(c.item.arguments),
        });
      } else if (eventType === "response.function_call_arguments.delta") {
        appendResponsesArguments(toolCallAcc, c, false);
      } else if (eventType === "response.function_call_arguments.done") {
        appendResponsesArguments(toolCallAcc, c, true);
      } else if (eventType === "response.completed" && isRecord(c.response)) {
        usage = usageFromWire(c.response.usage);
        const output = c.response.output;
        if (Array.isArray(output)) {
          for (const [idx, item] of output.entries()) {
            if (isRecord(item) && item.type === "function_call") {
              toolCallAcc.set(idx, {
                id: stringValue(item.call_id ?? item.id),
                name: stringValue(item.name),
                arguments: stringValue(item.arguments),
              });
            }
          }
        }
      } else if (eventType === "response.refusal.delta" && typeof c.delta === "string" && c.delta) {
        yield new ErrorChunk({ message: `Model refused: ${c.delta}` });
        return;
      }

      usage = usageFromWire(c.usage) ?? usage;

      const choices = c.choices as Record<string, unknown>[] | undefined;
      if (choices && choices.length > 0) {
        const delta = choices[0].delta as Record<string, unknown> | undefined;
        if (delta) {
          if (typeof delta.content === "string" && delta.content) {
            yield new TextChunk({ value: delta.content });
          }

          const tcDeltas = delta.tool_calls as Record<string, unknown>[] | undefined;
          if (tcDeltas) {
            for (const tcDelta of tcDeltas) {
              const idx = typeof tcDelta.index === "number" ? tcDelta.index : 0;
              const acc = toolCallAcc.get(idx) ?? { id: "", name: "", arguments: "" };
              if (tcDelta.id) acc.id = stringValue(tcDelta.id);
              const fn = tcDelta.function as Record<string, unknown> | undefined;
              if (fn) {
                if (fn.name) acc.name = stringValue(fn.name);
                if (fn.arguments) acc.arguments += stringValue(fn.arguments);
              }
              toolCallAcc.set(idx, acc);
            }
          }

          if (typeof delta.refusal === "string" && delta.refusal) {
            yield new ErrorChunk({ message: `Model refused: ${delta.refusal}` });
            return;
          }
        }
      }
    }
  } catch (error) {
    yield new ErrorChunk({
      message: error instanceof Error ? error.message : String(error),
    });
    return;
  }

  const sortedIndices = [...toolCallAcc.keys()].sort((a, b) => a - b);
  for (const idx of sortedIndices) {
    const tc = toolCallAcc.get(idx)!;
    yield ToolChunk.load({ kind: "tool", toolCall: tc });
  }
  if (usage) {
    yield new UsageChunk({ usage });
  }
}

async function* legacyStreamGenerator(
  chunks: AsyncIterable<StreamChunk>,
): AsyncGenerator<string | ToolCall> {
  for await (const chunk of chunks) {
    if (chunk instanceof TextChunk) {
      yield chunk.value;
    } else if (chunk instanceof ToolChunk) {
      yield chunk.toolCall;
    } else if (chunk instanceof ErrorChunk) {
      throw new Error(chunk.message);
    }
  }
}

function appendResponsesArguments(
  calls: Map<number, { id: string; name: string; arguments: string }>,
  chunk: Record<string, unknown>,
  replace: boolean,
): void {
  const callId = stringValue(chunk.call_id);
  const entry = [...calls.values()].find((call) => call.id === callId);
  if (!entry) return;
  const value = stringValue(replace ? chunk.arguments : chunk.delta);
  entry.arguments = replace ? value : entry.arguments + value;
}

function usageFromWire(value: unknown): InvocationUsage | undefined {
  if (!isRecord(value)) return undefined;
  const input = numberValue(value.input_tokens ?? value.prompt_tokens);
  const output = numberValue(value.output_tokens ?? value.completion_tokens);
  if (input === undefined && output === undefined) return undefined;
  return new InvocationUsage({
    inputTokens: input ?? 0,
    outputTokens: output ?? 0,
    totalTokens: numberValue(value.total_tokens) ?? (input ?? 0) + (output ?? 0),
  });
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function stringValue(value: unknown): string {
  return typeof value === "string" ? value : "";
}

function numberValue(value: unknown): number | undefined {
  return typeof value === "number" && Number.isFinite(value) ? value : undefined;
}

// ---------------------------------------------------------------------------
// Responses API processing
// ---------------------------------------------------------------------------

/**
 * Process a Responses API response.
 *
 * Extracts:
 * - Text content from `output_text` or output message items
 * - Function tool calls from `function_call` output items
 * - JSON-parsed content when outputs is present
 */
function processResponsesApi(
  agent: Prompty,
  response: Record<string, unknown>,
): unknown {
  const output = response.output as Record<string, unknown>[];

  // Collect function calls
  const funcCalls: ToolCall[] = [];
  for (const item of output) {
    if (item.type === "function_call") {
      funcCalls.push({
        id: (item.call_id ?? item.id ?? "") as string,
        name: item.name as string,
        arguments: item.arguments as string,
      });
    }
  }

  if (funcCalls.length > 0) {
    return funcCalls;
  }

  // Text content — use output_text convenience field
  const outputText = response.output_text as string | undefined;
  if (outputText !== undefined) {
    // Structured output — JSON parse when outputs schema exists
    if (agent.outputs && agent.outputs.length > 0) {
      try {
        return createStructuredResult(JSON.parse(outputText) as Record<string, unknown>, outputText);
      } catch {
        return outputText;
      }
    }
    return outputText;
  }

  // Fallback: extract from output message items
  const texts: string[] = [];
  for (const item of output) {
    if (item.type === "message") {
      const content = item.content as Record<string, unknown>[] | undefined;
      if (content) {
        for (const part of content) {
          if (part.type === "output_text" || part.type === "text") {
            texts.push(part.text as string);
          }
        }
      }
    }
  }

  if (texts.length > 0) {
    const text = texts.join("");
    if (agent.outputs && agent.outputs.length > 0) {
      try {
        return createStructuredResult(JSON.parse(text) as Record<string, unknown>, text);
      } catch {
        return text;
      }
    }
    return text;
  }

  return response;
}

// ---------------------------------------------------------------------------
// Non-streaming response processing
// ---------------------------------------------------------------------------

function processChatCompletion(
  agent: Prompty,
  response: Record<string, unknown>,
): unknown {
  const choices = response.choices as Record<string, unknown>[];
  if (!choices || choices.length === 0) return null;

  const choice = choices[0];
  const message = choice.message as Record<string, unknown>;
  if (!message) return null;

  // Tool calls
  const toolCalls = message.tool_calls as Record<string, unknown>[] | undefined;
  if (toolCalls && toolCalls.length > 0) {
    return toolCalls.map((tc): ToolCall => {
      const fn = tc.function as Record<string, unknown>;
      return {
        id: tc.id as string,
        name: fn.name as string,
        arguments: fn.arguments as string,
      };
    });
  }

  // Content
  const content = message.content as string | null;

  // Refusal — when content is null but refusal is present, return the refusal
  if (content === null) {
    const refusal = message.refusal as string | undefined;
    if (typeof refusal === "string") return refusal;
    return null;
  }

  // Structured output — JSON parse when outputs schema exists
  if (agent.outputs && agent.outputs.length > 0) {
    try {
      return createStructuredResult(JSON.parse(content) as Record<string, unknown>, content);
    } catch {
      return content;
    }
  }

  return content;
}

function processEmbedding(response: Record<string, unknown>): unknown {
  const data = response.data as Record<string, unknown>[];
  if (data.length === 1) {
    return (data[0] as Record<string, unknown>).embedding;
  }
  return data.map((d) => (d as Record<string, unknown>).embedding);
}

function processImage(response: Record<string, unknown>): unknown {
  const data = response.data as Record<string, unknown>[];
  if (data.length === 1) {
    return data[0].url ?? data[0].b64_json;
  }
  return data.map((d) => d.url ?? d.b64_json);
}
